### ISSUE TITLE
How to draw the diagram in the third row of Figure 1a.

### ISSUE THREAD

**Issue by zhangtsingtsi** [Necolizer/CHASE#4]
Thanks for your great work!
I found your article very inspiring for my task. 
So I'd like to ask how to plot the point cloud data after processing through the CHASE module?
I would be truly grateful if you could share some insights or example code with me.

**Comment by Necolizer** [comment]
Hi @zhangtsingtsi 

Good question. I list my steps to do that below:
1. Ensure that you have pretrained weights of CHASE(+some specified backbone). We provided some in the [checkpoint repo](https://huggingface.co/Necolizer/CHASE).
2. Use the following `CHASWrapper` class to load the checkpoint. Set `strict=False` in `model.load_state_dict()`. You can see this Wrapper class remove the steps used in training and also self.backbone, since we just want data after processing (inferring) through the CHASE module. Additionally, you should set the class args (e.g., in_channels) according to which checkpoint you wanna load.
```python
import torch
import torch.nn as nn
from einops import rearrange


class CHASWrapper(nn.Module):
    def __init__(self, in_channels=3, num_frame=64, num_point=25, pooling_seg=[1,1,1], num_entity=2, c1=64, c2=8):
        super(CHASWrapper, self).__init__()

        out_channel = num_frame * num_point * num_entity

        self.pooling_seg = pooling_seg
        self.seg = self.pooling_seg[0]*self.pooling_seg[1]*self.pooling_seg[2]
        self.seg_num_list = [(num_frame//self.pooling_seg[0]), (num_point//self.pooling_seg[1]), (num_entity//self.pooling_seg[2])]
        self.seg_num = self.seg_num_list[0] * self.seg_num_list[1] * self.seg_num_list[2]
        self.shift = nn.Sequential(
            nn.Conv3d(in_channels=in_channels, out_channels=c1, kernel_size=1),
            nn.AdaptiveAvgPool3d((self.pooling_seg[0], self.pooling_seg[1], self.pooling_seg[2])),
            nn.Conv3d(c1, c2, 1, bias=False),
            nn.ReLU(inplace=True),
            nn.Conv3d(c2, out_channel, 1, bias=False),
        )
    
    def forward(self, x):
        N, C, T, V, M = x.size()

        sf = self.shift(x).view(N, T*M*V, -1)
        tx = rearrange(x, 'n c t v m -> n c (t v m)', t=T, m=M, v=V).contiguous()
        sf = (tx @ sf.softmax(dim=1)).unsqueeze(-1).expand(-1, -1, self.seg, self.seg_num)
        sf = rearrange(sf, 'n c (T V M) (t v m) -> n c (T t) (V v) (M m)', 
                    T=self.pooling_seg[0], V=self.pooling_seg[1], M=self.pooling_seg[2], 
                    t=self.seg_num_list[0], v=self.seg_num_list[1], m=self.seg_num_list[2]).contiguous()
        x = x - sf

        return x
```
3. Load the dataset and feed the above `CHASWrapper` model by batch. Then you will get many batches of samples that only inferred by `CHASWrapper`. Visualize the skeleton joints of these samples using `matplotlib` to see the point cloud.

Hope the above steps might help.

**Comment by zhangtsingtsi** [comment]
> Hi [@zhangtsingtsi](https://github.com/zhangtsingtsi)
> 
> Good question. I list my steps to do that below:
> 
> 1. Ensure that you have pretrained weights of CHASE(+some specified backbone). We provided some in the [checkpoint repo](https://huggingface.co/Necolizer/CHASE).
> 2. Use the following `CHASWrapper` class to load the checkpoint. Set `strict=False` in `model.load_state_dict()`. You can see this Wrapper class remove the steps used in training and also self.backbone, since we just want data after processing (inferring) through the CHASE module. Additionally, you should set the class args (e.g., in_channels) according to which checkpoint you wanna load.
> 
> import torch
> import torch.nn as nn
> from einops import rearrange
> 
> 
> class CHASWrapper(nn.Module):
>     def __init__(self, in_channels=3, num_frame=64, num_point=25, pooling_seg=[1,1,1], num_entity=2, c1=64, c2=8):
>         super(CHASWrapper, self).__init__()
> 
>         out_channel = num_frame * num_point * num_entity
> 
>         self.pooling_seg = pooling_seg
>         self.seg = self.pooling_seg[0]*self.pooling_seg[1]*self.pooling_seg[2]
>         self.seg_num_list = [(num_frame//self.pooling_seg[0]), (num_point//self.pooling_seg[1]), (num_entity//self.pooling_seg[2])]
>         self.seg_num = self.seg_num_list[0] * self.seg_num_list[1] * self.seg_num_list[2]
>         self.shift = nn.Sequential(
>             nn.Conv3d(in_channels=in_channels, out_channels=c1, kernel_size=1),
>             nn.AdaptiveAvgPool3d((self.pooling_seg[0], self.pooling_seg[1], self.pooling_seg[2])),
>             nn.Conv3d(c1, c2, 1, bias=False),
>             nn.ReLU(inplace=True),
>             nn.Conv3d(c2, out_channel, 1, bias=False),
>         )
>     
>     def forward(self, x):
>         N, C, T, V, M = x.size()
> 
>         sf = self.shift(x).view(N, T*M*V, -1)
>         tx = rearrange(x, 'n c t v m -> n c (t v m)', t=T, m=M, v=V).contiguous()
>         sf = (tx @ sf.softmax(dim=1)).unsqueeze(-1).expand(-1, -1, self.seg, self.seg_num)
>         sf = rearrange(sf, 'n c (T V M) (t v m) -> n c (T t) (V v) (M m)', 
>                     T=self.pooling_seg[0], V=self.pooling_seg[1], M=self.pooling_seg[2], 
>                     t=self.seg_num_list[0], v=self.seg_num_list[1], m=self.seg_num_list[2]).contiguous()
>         x = x - sf
> 
>         return x
> 3. Load the dataset and feed the above `CHASWrapper` model by batch. Then you will get many batches of samples that only inferred by `CHASWrapper`. Visualize the skeleton joints of these samples using `matplotlib` to see the point cloud.
> 
> Hope the above steps might help.

Thank you so much! It‘s helpful. Wishing you all the best with your project!
